### PR TITLE
Enable S3 Artifact usage

### PR DESCRIPTION
**Pull request by Seanny123**
Currently, only OSS Artifacts are supported. This PR adds an `S3` Artifact sub-class.

I still haven't added tests for this, but I wanted to open this PR to discuss how this type of change is supposed to be implemented. Is the method in the first commit, where I duplicate code between `OSSArtifact` and `S3Artifact`, acceptable? Alternatively, is the second commit, where I create a `TypedArtifact` subclass better?

### CODE REVIEW

File: couler/core/templates/artifact.py [hunk 1]
```diff
@@ -37,30 +37,36 @@ def to_yaml(self):
...
     def __init__(
         self,
+        artifact_type,
         path,
         accesskey_id,
         accesskey_secret,
         bucket,
         key=None,
-        endpoint="http://oss-cn-hangzhou-zmf.aliyuncs.com",
+        endpoint="s3.amazonaws.com",
```
Comment (terrytangyuan) [inline couler/core/templates/artifact.py:56]: Remove this s3-specific default since it's the base class?

File: couler/core/templates/artifact.py [hunk 2]
```diff
@@ -79,7 +85,7 @@ def __init__(
         self.secret = couler.create_secret(secrets)
 
     def to_yaml(self):
-        oss_config = OrderedDict(
+        s3_config = OrderedDict(
```
Comment (terrytangyuan) [inline couler/core/templates/artifact.py:88]: Same here. This wouldn't be s3 specific then.

**Review by terrytangyuan — COMMENTED** [review]
Thanks for the contribution! IMO the second commit would look more concise. Just left some small comments. Otherwise once the test is in place then it should be good to merge.

**Comment by terrytangyuan** [comment]
@Seanny123 I noted that you removed the "WIP" label. Would you like to add some unit test to cover the modified code?

**Comment by Seanny123** [comment]
@terrytangyuan I'm having a hard time getting my tests to work. Does an actual `S3` resource have to exist or is it just mocked out?

**Review by terrytangyuan — COMMENTED** [review]
The unit tests only generates a YAML and you can get it via `couler.workflow_yaml` as a dictionary like what you are doing. They are not submitted to k8s. So you are definitely on the right track. Try checking whether the s3 related artifact spec exists in the dictionary.

**Comment by Seanny123** [comment]
Figured it out! Are the tests I wrote sufficient?

**Review by terrytangyuan — APPROVED** [review]
Looks good. Thanks!

**Pull request merged**